### PR TITLE
Nested arrays with multiple data types

### PR DESCRIPTION
**Pull request by bobbyrc**
### Summary:
When creating a schema with an attribute containing a nested array with multiple data types, Dynamoose incorrectly threw an error thinking that the array was actually a nested index.
This small fix checks to see if the nested property is an array before throwing an error about nested indexes.

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I have ensured the following commands are successful from the root of the project directory
  - [x] `npm test`
  - [x] `npm run lint`
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "dynamoose",
-  "version": "2.3.0",
+  "version": "2.3.1",
```
Comment (fishcharlie) [inline package.json:3]: This needs to be reverted. This (along with `package-lock.json` version) will be bumped automatically as part of our automated release process.

```suggestion
  "version": "2.3.0",
```

File: lib/Schema.ts [hunk 2]
```diff
@@ -582,7 +582,8 @@ export class Schema {
 		}
 
 		this.attributes().forEach((key) => {
-			if (key.includes(".") && this.getAttributeSettingValue("index", key)) {
+			const attributeSettingValue = this.getAttributeSettingValue("index", key);
+			if (key.includes(".") && attributeSettingValue && !Array.isArray(attributeSettingValue)) {
```
Comment (fishcharlie) [inline lib/Schema.ts:586]: So the problem with this is that in the event you have an index with multiple types, it won't throw an error. For example, the following schema should throw an error `Index must be at root object and not nested in object or array.`.

```js
new dynamoose.Schema({
    "id": String,
    "obj": {
        "type": Object,
        "schema": {
            "data": [
                {"type": Array, "schema": [String]},
                {"type": String, "index": true}
            ]
        }
    }
});
```

However, with the current code, it doesn't look like it'll throw an error. I think the solution is something like:

```suggestion
			const attributeSettingValue = this.getAttributeSettingValue("index", key);
			if (key.includes(".") && (Array.isArray(attributeSettingValue) ? attributeSettingValue.some((singleValue) => Boolean(singleValue)) : attributeSettingValue)) {
```

That way if it's an array, we are checking to see if `index` exists on any of those types.

File: test/unit/Schema.js [hunk 3]
```diff
@@ -76,6 +76,10 @@ describe("Schema", () => {
 		expect(() => new dynamoose.Schema({"id": String, "friend": {"type": Object, "schema": {"name": {"type": String, "index": {"global": true}}}}})).to.throw("Index must be at root object and not nested in object or array.");
 	});
 
+	it("Should not throw error if passing multiple data types for a nested array attribute", () => {
+		expect(() => new dynamoose.Schema({"id": String, "friends": {"type": Object, "schema": {"names": [{"type": Array, "schema": [String]}, {"type": Array, "schema": [Number]}]}}})).to.not.throw();
+	});
+
```
Comment (fishcharlie) [inline test/unit/Schema.js:82]: Above I gave another example of a case where it should fail. We should add that as a test, and make sure it throws an error. But we should make it fit this pattern (using the name `friends`, etc.).

We also need to confirm that that test fails without the change I mentioned above, and passes with the change I mentioned above.

File: test/unit/Schema.js [hunk 4]
```diff
@@ -76,6 +76,10 @@ describe("Schema", () => {
 		expect(() => new dynamoose.Schema({"id": String, "friend": {"type": Object, "schema": {"name": {"type": String, "index": {"global": true}}}}})).to.throw("Index must be at root object and not nested in object or array.");
 	});
 
+	it("Should not throw error if passing multiple data types for a nested array attribute", () => {
```
Comment (fishcharlie) [inline test/unit/Schema.js:79]: Just to confirm, this test fails without the changes you already made, correct?

File: package.json (hunk 1 above)
Comment (bobbyrc) [reply package.json:3]: Ah gotcha. Sorry if I overlooked that somewhere!

File: test/unit/Schema.js (hunk 4 above)
Comment (bobbyrc) [reply test/unit/Schema.js:79]: Yep, this test currently fails without the array check and passes with it.

File: lib/Schema.ts (hunk 2 above)
Comment (bobbyrc) [reply lib/Schema.ts]: For sure. Apologizes for missing this, it was difficult to jump into dynamoose and get my bearings quickly. There's a ton of complexity everywhere throughout the code. Impressed seeing the work as I was looking through.

File: test/unit/Schema.js (hunk 3 above)
Comment (bobbyrc) [reply test/unit/Schema.js:86]: Added a test to account for this case!

File: lib/Schema.ts (hunk 2 above)
Comment (fishcharlie) [reply lib/Schema.ts]: @bobbyrc No problem at all! It's not easy jumping into a new codebase. No need at all to apologize for that. Truly appreciate your help on this!! It only gets easier from here haha.

**Comment by bobbyrc** [comment]
Commenting to remove stale tag

**Comment by fishcharlie** [comment]
Made a lot of progress on this PR. Added some more tests which revealed some potential bugs. Currently it looks like Dynamoose is only respecting the first entry in the array of entries. So if in your schema you allow an array of strings and an array of numbers, and pass in an array of numbers, it will throw an error (when it should be accepting that).

The only thing to do to complete this PR is to fix it so those tests pass, and I wanna take one more scan of the code changes after we complete that.

---

I've kinda traced the problem to the following line:

https://github.com/dynamoose/dynamoose/blob/0202bdaa5dbc5558a7c363536896731e7da5e3b3/lib/Schema.ts#L405

In some cases, that is returning `false`, when it should be returning `true`. In other cases it correctly returns `false`.

Digging further. The `getValueTypeCheckResult` method, `key` is equal to `friends.names.1`. Which does not seem right. That causes `typeDetails` to be a string type.

image

In reality I expect this to be an array of array types. Like the following.

image

One *possible* solution would be to change the key from `${fullKey}.${subKey}` to `fullKey`. Problem with that, is I *think* that will cause a lot of regressions. Not positive on that tho.

I forget exactly how this logic all works, and all the different edge cases we need to account for.

I'm not sure why this is causing problems now either. All of this code was tested to work with multiple types. I think it could be the fact that we have an array we are passing in as the property value in the document, and an array of possible types. And when you do `friends.names.1` or `friends.names.0` or whatever, it's not sure if it should be referring to the value, or the type. But there are reasons why I doubt that. Like the fact that it's obvious what should be going on, and what it should be referring to. However it is possible that it's just a bug in the code that hasn't surfaced until now.

Point being, I'm not really sure at this point what is going on.

**Comment by fishcharlie** [comment]
Disregard what I said. This is fixed in 0b703d87d093193caecf17780b42bb445d17705f.

@bobbyrc Could you review (and possibly test) the changes I made and let me know what you think?

**Review by fishcharlie — APPROVED** [review]

**Review by fishcharlie — APPROVED** [review]

**Pull request merged**